File: go/micro/receiver.go
package main

import (
	"fmt"
	"net"
)

func main() {
	ln, err := net.Listen("tcp", ":9000")
	if err != nil {
		panic(err)
	}
	defer ln.Close()
	fmt.Println("Receiver listening on 127.0.0.1:9000")
	conn, err := ln.Accept()
	if err != nil {
		panic(err)
	}
	defer conn.Close()
	buffer := make([]byte, 0)
	tmp := make([]byte, 1024)
	for {
		n, err := conn.Read(tmp)
		if err != nil {
			break
		}
		buffer = append(buffer, tmp[:n]...)
		// 尝试解析完整帧
		for len(buffer) >= 5 {
			if buffer[0] != StartByte {
				buffer = buffer[1:]
				continue
			}
			length := buffer[2]
			frameLen := int(4 + length)
			if len(buffer) < frameLen {
				break
			}
			frame := buffer[:frameLen]
			cmd, data, err := UnpackFrame(frame)
			if err != nil {
				fmt.Println("Frame error:", err)
				buffer = buffer[1:]
				continue
			}
			fmt.Printf("Received: cmd=%d, data=%s\n", cmd, string(data))
			buffer = buffer[frameLen:]
		}
	}
}

File: mcp_micro_java/Receiver.java
import java.io.InputStream;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Arrays;

public class Receiver {
    public static void main(String[] args) throws Exception {
        try (ServerSocket server = new ServerSocket(9000)) {
            System.out.println("Receiver listening on 127.0.0.1:9000");
            try (Socket socket = server.accept()) {
                InputStream in = socket.getInputStream();
                byte[] buffer = new byte[1024];
                int bufferLen = 0;
                while (true) {
                    int n = in.read(buffer, bufferLen, buffer.length - bufferLen);
                    if (n == -1) break;
                    bufferLen += n;
                    // 尝试解析完整帧
                    while (bufferLen >= 5) {
                        if (buffer[0] != MCPProtocol.START_BYTE) {
                            // 跳过无效起始字节
                            System.arraycopy(buffer, 1, buffer, 0, --bufferLen);
                            continue;
                        }
                        int length = buffer[2] & 0xFF;
                        int frameLen = 4 + length;
                        if (bufferLen < frameLen) break;
                        byte[] frame = Arrays.copyOfRange(buffer, 0, frameLen);
                        try {
                            MCPProtocol.UnpackedFrame unpacked = MCPProtocol.unpackFrame(frame);
                            System.out.printf("Received: cmd=%d, data=%s\n", unpacked.cmd, new String(unpacked.data));
                        } catch (Exception e) {
                            System.out.println("Frame error: " + e.getMessage());
                        }
                        // 移动buffer
                        System.arraycopy(buffer, frameLen, buffer, 0, bufferLen - frameLen);
                        bufferLen -= frameLen;
                    }
                }
            }
        }
    }
} 